### ISSUE TITLE
Datasets

### ISSUE THREAD

**Issue by bali-eng** [pursuecong/WinGNN#5]
Hello, thank you for the great work.

I have a question regarding other datasets (dblp, bitcoinotc, etc.). In #2, you provided the link to the dblp dataset. Could you please clarify how you formatted it into a time series? Additionally, could you provide all your datasets for reproducibility purposes? I know you used ROLAND for preprocessing. ROLAND used the GraphDataset class to handle the data, while your code accepts data from the dgl.heterograph.DGLHeteroGraph class. It would be nice if it's possible to access your processed datasets.

Thanks

**Comment by maralgolabi** [comment]
Hi. Did you find how to use the dblp dataset? I need the three .npy files which are used in the load file and can't find them.

**Comment by pursuecong** [comment]
def create_dataset():
    ## Load dataset
    time1 = time.time()
    # 数据集是来自 pyg 还是 自己的
    if cfg.dataset.format in ['OGB']:
        graphs, splits = load_dataset()
    else:
        if cfg.dataset.format == 'dblp':
            dataset = 'dblp'
            e_feat = np.load('/share/share_40t/fangpeng/dataset/{0}/ml_{0}.npy'.format(dataset))
            n_feat = np.load('/share/share_40t/fangpeng/dataset/{0}/ml_{0}_node.npy'.format(dataset))
            e_time = np.load('/share/share_40t/fangpeng/dataset/{0}/ml_{0}_ts.npy'.format(dataset))
            train, train_e_feat, train_n_feat, test, test_e_feat, test_n_feat = load("Norandom", len(n_feat))
            n_feat = torch.Tensor(n_feat)
            e_feat = torch.Tensor(e_feat[:-1])
            # 将数据集全部组合起来
            graphs = []
            count = 0
            for tr in train:
                row = tr.row
                col = tr.col
                edge_index = np.vstack((row, col))
                edge_index = torch.Tensor(edge_index).long()
                e_time = [e_time[count] for i in range(edge_index.shape[1])]
                e_time = torch.Tensor(e_time)
                graphs.append(Graph(
                    node_feature=n_feat,
                    edge_feature=e_feat,
                    edge_index=edge_index,
                    edge_time=e_time,
                    directed=True
                ))
                count += 1

            for te in test:
                row = te.row
                col = te.col
                edge_index = np.vstack((row, col))
                edge_index = torch.Tensor(edge_index).long()
                e_time = [e_time[count] for i in range(edge_index.shape[1])]
                e_time = torch.Tensor(e_time)
                graphs.append(Graph(
                    node_feature=n_feat,
                    edge_feature=e_feat,
                    edge_index=edge_index,
                    edge_time=e_time,
                    directed=True
                ))
                count += 1
            for g_snapshot in graphs:
                g_snapshot.node_states = [0 for _ in range(cfg.gnn.layers_mp)]
                g_snapshot.node_cells = [0 for _ in range(cfg.gnn.layers_mp)]
                g_snapshot.node_degree_existing = torch.zeros(n_feat.size(0))
        else:
            graphs = load_dataset()



    ## Filter graphs
    time2 = time.time()
    # n = math.ceil(len(graphs) * 0.7)
    # test_data = graphs[n-1:]
    # graphs = graphs[:n]
    min_node = filter_graphs()

    # 保存数据
    # for i, graph in enumerate(graphs):
    #     edge_index = graph.edge_index.numpy()
    #     edge_feature = graph.edge_feature.numpy()
    #     node_feature = graph.node_feature.numpy()
    #     edge_time = graph.edge_time.numpy()
    #
    #     path = "/share/share_40t/fangpeng/dataset/stackoverflow_M"
    #
    #     if not os.path.exists(path + '/' + 'edge_index'):
    #         os.makedirs(path + '/' + 'edge_index')
    #     path_ei = path + '/' + 'edge_index/' + str(i) + '.npy'
    #
    #     if not os.path.exists(path + '/' + 'node_feature'):
    #         os.makedirs(path + '/' + 'node_feature')
    #     path_nf = path + '/' + 'node_feature/' + str(i) + '.npy'
    #
    #     if not os.path.exists(path + '/' + 'edge_feature'):
    #         os.makedirs(path + '/' + 'edge_feature')
    #     path_ef = path + '/' + 'edge_feature/' + str(i) + '.npy'
    #
    #     if not os.path.exists(path + '/' + 'edge_time'):
    #         os.makedirs(path + '/' + 'edge_time')
    #     path_et = path + '/' + 'edge_time/' + str(i) + '.npy'
    #
    #     np.save(path_ei, edge_index)
    #     np.save(path_ef, edge_feature)
    #     np.save(path_nf, node_feature)
    #     np.save(path_et, edge_time)



    ## Create whole dataset

    dataset = GraphDataset(
        graphs,
        task=cfg.dataset.task,
        edge_train_mode=cfg.dataset.edge_train_mode,
        edge_message_ratio=cfg.dataset.edge_message_ratio,
        edge_negative_sampling_ratio=cfg.dataset.edge_negative_sampling_ratio,
        minimum_node_per_graph=min_node)
    # graphs = dataset.graphs[0].edge_label_index
    # graphs = graphs[:, graphs.size(1) // 2:]

    # test_data = GraphDataset(
    #     test_data,
    #     task=cfg.dataset.task,
    #     edge_train_mode=cfg.dataset.edge_train_mode,
    #     edge_message_ratio=cfg.dataset.edge_message_ratio,
    #     edge_negative_sampling_ratio=cfg.dataset.edge_negative_sampling_ratio,
    #     minimum_node_per_graph=min_node)

    ## Transform the whole dataset
    dataset = transform_before_split(dataset)
    # test_data = transform_before_split(test_data)

    ## Split dataset
    time3 = time.time()
    # Use custom data splits
    if cfg.dataset.split_method == 'chronological_temporal':
        if cfg.train.mode == 'live_update_fixed_split':
            datasets = [dataset, dataset, dataset]
        else:
            total = len(dataset)  # total number of snapshots.
            train_end = int(total * cfg.dataset.split[0])
            val_end = int(total * (cfg.dataset.split[0] + cfg.dataset.split[1]))
            datasets = [
                dataset[:train_end],
                dataset[train_end:val_end],
                dataset[val_end:]
            ]
    else:
        if cfg.dataset.format == 'OGB':
            datasets = []
            datasets.append(dataset[splits['train']])
            datasets.append(dataset[splits['valid']])
            datasets.append(dataset[splits['test']])
        # Use random split, supported by DeepSNAP
        else:

            # dataset = dataset.split(split_ratio=[0.7, 0.3], shuffle=False)
            datasets = dataset.split(
                transductive=cfg.dataset.transductive,
                split_ratio=cfg.dataset.split,
                shuffle=cfg.dataset.shuffle)
    ## Transform each split dataset
    time4 = time.time()
    datasets = transform_after_split(datasets)
    # test_data = transform_after_split(test_data)

    time5 = time.time()
    logging.info('Load: {:.4}s, Before split: {:.4}s, '
                 'Split: {:.4}s, After split: {:.4}s'.format(
                 time2 - time1, time3 - time2, time4 - time3, time5 - time4))
    # return datasets, test_data
    return datasets****
Replace the create_dataset function in loader.py in graphgym in roland code

**Comment by pursuecong** [comment]
Thank you for your attention to our work

**Comment by bali-eng** [comment]
@pursuecong Thank you for your reply. I assume the YAML file is required for the Roland code, and the data itself should also be available. Thanks for sharing how you prepared the data.
@maralgolabi I found those three files and created a custom function to generate the dataset.

**Comment by maralgolabi** [comment]
@bali-eng Can you tell me where you found them? Thank you for your response.

**Comment by bali-eng** [comment]
@maralgolabi  I got them from a colleague. 
But to give you some hints, you can download the text file from [the second link ](https://github.com/BUPT-GAMMA/MetaDyGNN/tree/main/data/dblp). Then, use [this script](https://github.com/BUPT-GAMMA/MetaDyGNN/blob/main/code/process_dblp.py) to process the data. Note that the initial node features are set to 0.

This might also be helpful: [PyTorch Geometric DBLP dataset](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/datasets/dblp.py).

**Comment by maralgolabi** [comment]
@bali-eng Great! Thank you! The only problem I'm facing is the third npy file (the ts one) which I don't know yet how to work out. Can you please give me a hint on that too?

**Comment by bali-eng** [comment]
@maralgolabi sure. Checking [this script](https://github.com/BUPT-GAMMA/MetaDyGNN/blob/main/code/process_dblp.py), it seems the last column in the .txt file is time. @pursuecong can confirm this.
This line specifies the path to the .txt file: PATH = '../data/{0}/{0}.txt'.format(data_name).
The preprocess function (df = preprocess(PATH)) treats the first column as the source node, the second as the destination node, and the third as the timestamp (ts). 


If you need edge indexes, you can use source and destination nodes.
